src/dispatching/update_listeners/polling.rs

```diff
@@ -123,131 +226,179 @@ where
 /// ```
 ///
 /// [get_updates]: crate::requests::Requester::get_updates
-pub fn polling<R>(
-    bot: R,
+/// [`Dispatcher`]: crate::dispatching::Dispatcher
+pub struct Polling<B: Requester> {
+    bot: B,
     timeout: Option<Duration>,
     limit: Option<u8>,
     allowed_updates: Option<Vec<AllowedUpdate>>,
-) -> impl UpdateListener<R::Err>
+    drop_pending_updates: bool,
+    flag: AsyncStopFlag,
+    token: AsyncStopToken,
+}
+
+impl<R> Polling<R>
 where
     R: Requester + Send + 'static,
     <R as Requester>::GetUpdates: Send,
 {
-    struct State<B: Requester> {
-        bot: B,
-        timeout: Option<u32>,
-        limit: Option<u8>,
-        allowed_updates: Option<Vec<AllowedUpdate>>,
-        offset: i32,
-        flag: AsyncStopFlag,
-        token: AsyncStopToken,
-        force_stop: bool,
+    /// Returns a builder for polling update listener.
+    pub fn builder(bot: R) -> PollingBuilder<R> {
+        PollingBuilder {
+            bot,
+            timeout: None,
+            limit: None,
+            allowed_updates: None,
+            drop_pending_updates: false,
+        }
     }
+}
 
-    fn stream<B>(st: &mut State<B>) -> impl Stream<Item = Result<Update, B::Err>> + Send + '_
-    where
-        B: Requester + Send,
-        <B as Requester>::GetUpdates: Send,
-    {
-        stream::unfold(st, move |state| async move {
-            let State { timeout, limit, allowed_updates, bot, offset, flag, force_stop, .. } =
-                &mut *state;
-
-            if *force_stop {
-                return None;
-            }
+#[pin_project::pin_project]
+pub struct PollingStream<'a, B: Requester> {
+    /// Parent structure
+    polling: &'a mut Polling<B>,
```

Comment:
We don't really need exclusive/unique reference here. A shared `&'a Polling<B>` would suffice. But, shared reference would require `B: Sync` instead of `B: Send`.

So the questions are:
1. Should I change this to a shared reference? (`&'a Polling<B>`)
2. Should I change bounds to be `Send + Sync`, so we can later change this however we want, if need be

Reply:
For the record: adding `Sync` bound would be a breaking change.

Reply:
Well, I think it's a tiny breaking change that will affect almost no one, but it would allow us to twiddle with the implementation later without BC. GIven that the next release is non-BC, I think we should change this to a shared reference and change the bounds to `Send + Sync`.

Reply:
Actually, we already have `B: Send` in `polling_default`, so adding `Sync` would be a BC.